nexus/src/app/background/instance_watcher.rs

```diff
@@ -0,0 +1,635 @@
+    }
+
+    impl Instance {
+        pub(super) fn success(&mut self, updated: InstanceUpdated) {
+            match updated {
+                InstanceUpdated {
+                    instance_updated: true,
+                    vmm_updated: true,
+                } => self.both_updated.datum += 1,
+                InstanceUpdated {
+                    instance_updated: true,
+                    vmm_updated: false,
+                } => self.instance_updated.datum += 1,
+                InstanceUpdated {
+                    instance_updated: false,
+                    vmm_updated: true,
+                } => self.vmm_updated.datum += 1,
+                InstanceUpdated {
+                    instance_updated: false,
+                    vmm_updated: false,
+                } => self.no_update.datum += 1,
+            }
+            self.touched = true;
+        }
+
+        pub(super) fn check_failure(&mut self, reason: CheckFailure) {
+            self.check_failures
+                .entry(reason)
+                .or_insert_with(|| InstanceCheckFailures {
+                    reason: reason.to_string(),
+                    datum: Cumulative::default(),
+                })
+                .datum += 1;
+            self.touched = true;
+        }
+
+        pub(super) fn check_error(&mut self, reason: CheckError) {
+            self.update_failures
+                .entry(reason)
+                .or_insert_with(|| InstanceCheckErrors {
+                    reason: reason.to_string(),
+                    datum: Cumulative::default(),
+                })
+                .datum += 1;
+            self.touched = true;
+        }
+
+        fn len(&self) -> usize {
+            4 + self.check_failures.len() + self.update_failures.len()
+        }
+
+        fn sample_into(
+            &self,
+            target: &InstanceTarget,
+            dest: &mut Vec<Sample>,
+        ) -> Result<(), MetricsError> {
+            dest.push(Sample::new(target, &self.no_update)?);
+            dest.push(Sample::new(target, &self.instance_updated)?);
+            dest.push(Sample::new(target, &self.vmm_updated)?);
+            dest.push(Sample::new(target, &self.both_updated)?);
+            for metric in self.check_failures.values() {
+                dest.push(Sample::new(target, metric)?);
+            }
+            for metric in self.update_failures.values() {
+                dest.push(Sample::new(target, metric)?);
+            }
+            Ok(())
+        }
+    }
+
+    /// The number of successful checks for a single instance and sled agent
+    /// pair.
+    #[derive(Clone, Debug, Metric)]
+    struct InstanceChecks {
+        /// `true` if the instance state changed as a result of this check.
+        instance_updated: bool,
+        /// `true` if the VMM state changed as a result of this check.
+        vmm_updated: bool,
+        /// The number of successful checks for this instance and sled agent.
+        datum: Cumulative<u64>,
+    }
```

Comment:
This metric structure seems a bit funky to me. I'm not sure I can put my finger on it just yet, so bear with me as I talk / write through this a bit.

As it stands, one would view this through the timeseries `instance_target:instance_checks`. The datum of that timeseries would count the total number of successful checks, with different timeseries for those where the instance or VMM was updated. (So 4 total timeseries.)

It seems like one might want to instead report the _state_ of the instance and VMM itself as part of the timeseries, or more concretely the count of times we successfully observed each state. This mirrors for example the vCPU usage, which reports the cumulative time in each of the tracked vCPU microstates. That way, I can ask questions like, "When was the instance running?", "How long was the instance stopped on this day?", or "Which instances are currently failed?".

I'm not sure how one would answer those questions with the knowledge of state _changes_ only. Also, if one really wants to know that, we can build methods for finding those points of change from the raw data itself. (Such a thing doesn't yet exist in OxQL, but I can squint and see how one might implement it.)

Concretely, I think we might prefer something like this:

```rust
#[derive(oximeter::Metric)]
struct InstanceState {
    // Stringified form of the same enum we report in the API
    state: String,
    // Count of times we observed this state
    datum: Cumulative<u64>,
}

#[derive(oximeter::Metric)]
struct VmmState {
    // Stringified form of the same enum we track internally about the VMM state
    state: String,
    // Count of times we observed this state
    datum: Cumulative<u64>,
}
```

Now that I think about it, a lot of this is assuming that this is _the place_ we report instance state. It's possible that we want Propolis to report these states directly, alongside the vCPU metrics it already publishes. If that is the intention, then most of what I've written here isn't relevant! I am still unclear on how one might use the boolean fields around the instance / VMM state changes, but it's much less important if the example questions I wrote earlier can be answered from another timeseries!

Reply:
I think labeling the timeseries with the actual observed state of the instance and the VMM seems like a good idea! I'll change this code to do that.

Reply:
@bnaecker if we were to start recording metrics for instance and VMM states, would it make more sense to:

1. add the VMM's UUID as a field on the `VirtualMachine` target, so the queries would be `virtual_machine:instance_state` and `virtual_machine:vmm_state`, where the `VirtualMachine` target includes a VMM UUID (and thus would not be equal if an instance migrated from one VMM to another), or,
2. create separate `VirtualMachine` and `VirtualMachineManager` targets, where only the `VirtualMachineManager` target has the VMM UUID?

Really, I guess, what I'm asking is, is it possible to query a target based on only one of its fields (such as the instance ID)? If not, the current structure might not make sense.